README

This is Net::SSH::Perl. It contains implementations of
both the SSH1 and SSH2 protocols.

PREREQUISITES

Protocol 1

    * Math::GMP (1.04 or greater)
    * String::CRC32 (1.2 or greater)
    * Digest::MD5
    * IO::Socket

Protocol 2

    * BSD::arc4random (1.50 or greater)
    * Crypt::DSA (0.03 or greater)
    * Crypt::DH (0.01 or greater)
    * Crypt::Curve25519 (0.04 or greater)
    * Crypt::Ed25519
    * Crypt::OpenSSL::AES
    * Math::Pari (2.001804 or greater)
    * MIME::Base64
    * Digest::MD5
    * Digest::SHA
    * Digest::HMAC_MD5
    * Convert::PEM (0.05 or greater)
    * IO::Socket

The module Crypt::OpenSSH::ChachaPoly is required if
you wish to use the chacha20-poly1305@openssh.com cipher.
It can be acquired from Git Hub at:
https://github.com/lkinley/Crypt-OpenSSH-ChachaPoly
or using git:
git clone https://github.com/lkinley/Crypt-OpenSSH-ChachaPoly.git
If the module is not installed, Net::SSH::Perl will fall back on
other ciphers.

The module Crypt::OpenBSD::Blowfish is required for encrypted
Ed25519 key support.  It can be acquired from Git Hub at:
https://github.com/lkinley/Crypt-OpenBSD-Blowfish
or using git:
git clone https://github.com/lkinley/Crypt-OpenBSD-Blowfish.git
If the module is not installed, only unencrypted Ed25519 keys
can be used.

It also optionally requires Digest::BubbleBabble for
generating bubble babble fingerprints, and Crypt::RSA
(1.37 or greater) if you want to use RSA key files in SSH2.

Note that RSA key files in SSH1 do not require Crypt::RSA.

INSTALLATION

Net::SSH::Perl installation is relatively straightforward. If
your CPAN shell is set up, you should just be able to do

    % perl -MCPAN -e 'install Net::SSH::Perl'

If you don't like that, you can download the distribution; the
latest version on CPAN can be found at

	http://search.cpan.org/dist/Net-SSH-Perl/

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

If you have problems, read on.

The only slightly complicated bit in the installation is
that you'll need to install Crypt:: modules depending on
which ciphers you wish to use. Much work has gone into
making this simple; you should be asked a question about
which ciphers you'll use, and the modules should be
downloaded and installed automatically.

If it doesn't work, or if you don't like this option, you'll
need to do the installations manually. In which case you'll
need to install the prerequisites listed above for your
chosen protocol, plus any additional Crypt:: modules you
wish to use.

SAMPLES/TUTORIALS

Take a look at the scripts in eg/ for help and examples of
using Net::SSH::Perl.

* eg/cmd.pl is just a simple example of some of the
functionality in the module;

# eg/pssh is an ssh-like client for running commands on
remote servers and connecting to a shell on remote servers;

* eg/pscp is a very simple scp-like script;

* eg/pssh-keygen is, as you might guess, an implementation of
ssh-keygen in Perl, using the Net::SSH::Perl key libraries.
It contains almost all of the functionality in the OpenSSH
version of ssh-keygen.

* eg/remoteinteract.pl is an example of using register_handler
to interact with a remote (interactive) command.

Both pssh and pscp support a subset of the command line options
that the actual tools support; obviously, only those options
supported by Net::SSH::Perl are supported by the two scripts.

David Robins / dbrobins@cpan.org

Support for the following features by:
Lance Kinley / lkinley@loyaltymethods.com

Diffie-Hellman Group 14 Key Exchange
Diffie-Hellman Group Exchange
Curve25519 Key Exchange (curve25519-sha256@libssh.org)
AES Cipher support (aes256-ctr,aes192-ctr,aes128-ctr)
Chacha20-Poly1305 cipher support (chacha20-poly1305@openssh.com)
hmac-sha2-256,hmac-sha2-512 MAC support
hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com
  Encrypt-then-MAC (ETM) MAC support
Ed25519 key support, including encrypted keys that are enciphered
  with supported crypto algorithms


Changes
Revision history for Net::SSH::Perl

2.02 2016.04.29

    - require Crypt::OpenSSH::ChachaPoly
    - require Crypt::OpenBSD::Blowfish
    - require Crypt::Curve25519
    - require Crypt::Ed25519

2.01 2016.02.19

    - Use CryptX to reduce module depedencies
      This eliminates the need for:
        BSD::arc4random
        Digest::MD5
        Digest::SHA
        Digest::HMAC_MD5
	Crypt::OpenSSL::AES

2.00 2015.12.07

    - Add Chacha20-Poly1305 cipher support for best security
      (Requires Crypt::OpenSSH::ChachaPoly, see README)
    - Add AES Cipher support in CTR mode (CBC mode supported in Ed25519
      keys only)
    - Add Group Exchange (RFC4523) Diffie-Hellman Key Exchange
    - Add Curve25519 (curve25519-sha256@libssh.org) Key Exchange support
      (Requires Crypt::Curve25519)
    - Add hmac-sha2-256,hmac-sha2-512 MAC support
    - Add hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com
      Encrypt-then-MAC (ETM) MAC support
    - Use BSD::arc4random for encrypted packet padding
    - Add support for Ed25519 ssh/host keys (Requires Crypt::Ed25519)
      Encrypted Ed25519 key support requires Crypt::OpenBSD::Blowfish
      (See README for info)
    - Default ciphers order is now chacha,aes,3des,blowfish,arcfour
    - Default KEX order is now Curve25519, DHGEXSHA256, DHGEXSHA1, DH14, DH1
    - Default MAC order is now hmac-sha2-512-etm@openssh.com,
      hmac-sha2-256-etm@openssh.com, sha2-512, sha2-256, sha1, md5
    - SSH Keys can now be in DOS format (no need to remove CR/LF)
    - SOCKS proxy support via sub class Net::SSH:Perl::Proxy
    - Now does not abort due to OpenSSH 6.8+ server
      SSH2_MSG_GLOBAL_REQUEST messages for host key rotation

1.42 2015.09.28
    - Fix issue with long selects getting interrupted by signals and dying (Andrew Hoying)
    - fix version cpan meta info

1.41 2015.09.18
    - use Errno constants in more portable way, see perldoc Errno

1.40 2015.09.17
    - declare new dependency to File::HomeDir in Makefile.PL

1.39 2015.09.15

    - RT #83978 - fix shell terminal width and height (lharey)
    - RT #94574 - fix Algorithm negotiation issue in ::Kex.pm
      (Brian Curnow, Michael Gray)
    - RT #105728 - fix VERSION methods (reported by Karen Etheridge)
    - code modernization (strict+warnings) (gh0stwizard)
    - pass tests on Win32 (gh0stwizard)
    - use Win32::LoginName on Windows (Michael Gray)
    - use File::HomeDir to simplify handling (Michael Gray)

1.38 2014.10.06

    - RT #99284 - install valid SIGNATURE file (Greg Sabino Mullane)

1.37 2014.03.17

    - RT #91840 - enabled config option "StrictHostKeyChecking"
                  (the corresponding code already existed)

1.36 2013.08.09

  Apply many bugfixes from RT discussions.

    - RT #48338 - FIX race condition with SSHv2
    - RT #55195 - FIX race condition in KEXINIT
    - RT #67586 - FIX test '03-packet.t' hangs forever
    - RT #64517 - enable PTY support in SSH2
    - RT #23947 - Replacement for KeyboardInt.pm

1.35 2012.12.03

    - rt#76482 - apply patch to t/03-packet.t for 5.15+ (chisel++)

1.34 2009.02.01

    - Rekey properly after 1 GB of data (rt.cpan.org #25044). Patch by Peter Oliver.
    - Don't try to process nonexistent or empty auth file (rt.cpan.org #41877).
    - Fix typo in croak message (rt.cpan.org #42056), thanks to jamie at audible.transient.net.
    - Move 'use base' call after Crypt module loading, per suggestion (rt.cpan.org #42051).
    - Only apply stdin if defined in SSH1 - John Payne (rt.cpan.org #42583)

1.33 2008.10.21

    - Fix open() calls (rt.cpan.org #40020)
    - Fix non-shell problem (rt.cpan.org #39980)
    - Allow full agent forwarding (rt.cpan.org #32190)
    - Handle hashed known_hosts files (Greg Sabino Mullane, rt.cpan.org #25175)

1.32 2008.10.16

    - Add IO::Handle to Perl.pm (rt.cpan.org #40057, #35985)
    - Minor test cleanups.

1.31 2008.10.02

    - New co-maintainer, Greg Sabino Mullane (TURNSTEP).
    - Prevent t/03-packet.t from hanging due to high file descriptor.
      (altblue at n0i.net, rt.cpan.org #6101)
    - Skip some tests if Math::GMP not installed (e.g. from choosing only 
      protocol 2 in Makefile.PL) (Greg Sabino Mullane, reported in 
      rt.cpan.org #25152)
    - If ENV{HOME} is not set, use getpwuid. If both fail and the dir 
      is needed, we croak. (Greg Sabino Mullane, expanded from patch 
      by dgehl at inverse.ca in rt.cpan.org #25174)
    - Fix incorrect logical/bitwise AND mixup (Peter.Haydon at uk.fujitsu.com, 
      rt.cpan.org #31490)
    - Allow empty stdin for SSH2 (rcp at rcable.co.uk, rt.cpan.org #32730)
    - Adjust terminal dimensions dynamically if Term::ReadKey is available
      (john at sackheads.org, rt.cpan.org #34874)

1.30  2005.10.12
    - Fix for local *READ/*WRITE tie problem in open2 function (Bas van
      Sisseren).

1.29  2005.10.08
    - Require Digest::SHA1 2.10 or greater (fixed problem with BSD connection)
      (Adam Callis).
    - Add suppress_shell SSH2 option for SFTP (Sean McMillan).
    - Use effective (not real) user-id to detect root (David Wheeler,
      rt.cpan.org #13373).
    - Make "ChallengeResponseAuthentication" configuration item boolean rather
      than string (mark AT blackmans.org, rt.cpan.org #14549).

1.28  2005.03.17
    - Patch to skip warning messages emitted by F-Secure sshd before the SSH
      signature (Neal Ley).
    - Fix for "Can't locate object method "blocking" via package "IO::Handle""
      error (rt.cpan.org #11674, Daniel C. Silverstein (dans AT pch.net)).

1.27  2005.02.06
    - Set socket to non-blocking in more portable manner (rt.cpan.org #6728).
    - Default port to 22 in case service list is not available (rt.cpan.org
      #5924, Gavin Brock).
    - Allow Net::SSH::Ciphers::supported to take a protocol option to get
      SSH2 ciphers (rt.cpan.org #4648, dan at midstream.com), and also change
      Net::SSH::Perl::Kex to map SSH1 names to SSH2 names.

Makefile.PL
# $Id: Makefile.PL,v 1.30 2008/10/02 18:53:25 turnstep Exp $

use strict;
use 5.006;
use ExtUtils::MakeMaker qw(prompt WriteMakefile);

my %prereq = (
        'IO::Socket'        => 0,
        'File::Spec'        => 0,
        'File::HomeDir'     => 0,
);

my %SSH_PREREQ = (
    1 => {
        'Digest::MD5'       => 0,
        'String::CRC32'     => '1.2',
        'Math::GMP'         => '1.04',
        'Scalar::Util'      => 0,
    },

    2 => {
        'Crypt::OpenSSH::ChachaPoly' => '0.02',
        'Crypt::OpenBSD::Blowfish' => '0.01',
        'Crypt::Curve25519' => '0.05',
        'Crypt::Ed25519'    => '1.03',
        'CryptX'            => 0,
        'Crypt::Curve25519' => '0.05',
        'Crypt::DH'         => '0.01',
        'Math::Pari'        => '2.001804',
        'MIME::Base64'      => 0,
        'Convert::PEM'      => '0.05',
    },
);

$SSH_PREREQ{3} = { map %{$SSH_PREREQ{$_}}, 1..2 };

print<<MSG;
This is Net::SSH::Perl.

As of version 1.00, Net::SSH::Perl supports both the SSH1 and
SSH2 protocols natively. The two protocols have different
module prerequisitives, so you need to decide which protocol(s)
you plan to use. If you use one or the other, only those modules
for your chosen protocol will be installed; if you choose both,
all of the supporting modules will be installed. Please choose
the protocols you'd like to use from the following list ("Both"
is the default).

MSG

printf "    [%d] SSH%d\n", $_, $_ for 1..2;
printf "    [3] Both SSH1 and SSH2\n";

my $p = prompt("\nWhich protocol(s) do you plan to use?", 3);
print "\n";

@prereq{keys %{$SSH_PREREQ{$p}}} = values %{$SSH_PREREQ{$p}};

print "Checking for optional modules\n\n";

unless (have_module('Digest::BubbleBabble', 0.01)) {
    print<<MSG, "\n";
Digest::BubbleBabble is required if you want to generate bubble babble
key fingerprints with pssh-keygen.
MSG

    if (read_yes_or_no("Would you like to install it? (y/n)", "y")) {
        $prereq{'Digest::BubbleBabble'} = '0.01';
    }
    print "\n";
}

WriteMakefile(
    (MM->can('signature_target') ? (SIGN => 1) : ()),
    NAME         => 'Net::SSH::Perl',
    DISTNAME     => 'Net-SSH-Perl',
    VERSION_FROM => 'lib/Net/SSH/Perl.pm',
    PREREQ_PM    => \%prereq,
    AUTHOR       => 'David Robins <dbrobins@cpan.org>',
    ABSTRACT     => 'Perl client interface to SSH',
    NO_META      => 1,
);

sub read_yes_or_no {
    my($prompt, $def) = @_;
    my $ans = prompt($prompt, $def);
    $ans =~ /^y/i;
}

sub have_module {
    my($name, $ver) = @_;
    eval("use $name" . ($ver ? " $ver;" : ";"));
    !$@;
}
